synthesizer/process/src/finalize.rs

```diff
@@ -187,80 +178,236 @@ fn finalize_transition<N: Network, P: FinalizeStorage<N>>(
+                        // If the evaluation fails, bail and return the error.
+                        Err(_) => bail!("'finalize' failed to evaluate command ({command})"),
+                    }
                     counter += 1;
-                    operations
                 }
-            }));
-
-            match result {
-                // If the evaluation succeeds with an operation, add it to the list.
-                Ok(Ok(Some(finalize_operation))) => finalize_operations.push(finalize_operation),
-                // If the evaluation succeeds with no operation, continue.
-                Ok(Ok(None)) => (),
-                // If the evaluation fails, bail and return the error.
-                Ok(Err(error)) => bail!("'finalize' failed to evaluate command ({command}): {error}"),
-                // If the evaluation fails, bail and return the error.
-                Err(_) => bail!("'finalize' failed to evaluate command ({command})"),
-            }
+            };
         }
     }
+
     // Return the finalize operations.
     Ok(finalize_operations)
 }
 
+// A helper struct to track the execution of a finalize block.
+struct FinalizeState<'a, N: Network> {
+    // A counter for the index of the commands.
+    counter: usize,
+    // The finalize logic.
+    finalize: &'a Finalize<N>,
+    // The registers.
+    registers: FinalizeRegisters<N>,
+    // The stack.
+    stack: &'a Stack<N>,
+    // Call counter.
+    call_counter: usize,
+    // Recent call register.
+    recent_call_locator: Option<u64>,
+}
+
+// A helper function to initialize the finalize state.
+fn initialize_finalize_state<'a, N: Network>(
+    state: FinalizeGlobalState,
+    future: &Future<N>,
+    stack: &'a Stack<N>,
+    transition_id: N::TransitionID,
+) -> Result<FinalizeState<'a, N>> {
+    // Get the finalize logic and the stack.
+    let (finalize, stack) = match stack.program_id() == future.program_id() {
+        true => (stack.get_function_ref(future.function_name())?.finalize_logic(), stack),
+        false => {
+            let stack = stack.get_external_stack(future.program_id())?;
+            (stack.get_function_ref(future.function_name())?.finalize_logic(), stack)
+        }
+    };
+    // Check that the finalize logic exists.
+    let finalize = match finalize {
+        Some(finalize) => finalize,
+        None => bail!(
+            "The function '{}/{}' does not have an associated finalize block",
+            future.program_id(),
+            future.function_name()
+        ),
+    };
+    // Initialize the registers.
+    let mut registers = FinalizeRegisters::new(
+        state,
+        transition_id,
+        *future.function_name(),
+        stack.get_finalize_types(future.function_name())?.clone(),
+    );
+
+    // Store the inputs.
+    finalize.inputs().iter().map(|i| i.register()).zip_eq(future.arguments().iter()).try_for_each(
```

Comment:
Need to add a length check for `finalize.inputs()` and `future.arguments()` before this line so the `zip_eq` doesn't panic.